### PR TITLE
Distributed Training

### PR DESCRIPTION
**Pull request by ajhinsvark**
Allow multi-gpu training for curriculum learning by creating a custom lazy dataloader for each task, instead of using pytorch's `utils.data.DataLoader`.

### CODE REVIEW

File: espnet2/curriculum/curriculum_iter_factory.py [hunk 1]
```diff
@@ -56,30 +83,24 @@ def build_iter(self, epoch=1):
         for i in range(len(self.sampler)):
             random.shuffle(self.sampler[i])
             self.loaders.append(
-                DataLoader(
-                    dataset=self.dataset,
-                    batch_sampler=self.sampler[i],
-                    num_workers=self.num_workers,
-                    pin_memory=self.pin_memory,
```
Comment (ajhinsvark) [inline espnet2/curriculum/curriculum_iter_factory.py:63]: num_workers and pin memory are ignored in the new TaskDataLoader.

File: espnet2/curriculum/curriculum_sampler.py [hunk 2]
```diff
@@ -64,14 +64,15 @@ def __init__(
         tasks = load_num_sequence_text(task_file, loader_type="text_int")
 
         first_utt2shape = utt2shapes[0]
+        first_keys = set(first_utt2shape)
```
Comment (ajhinsvark) [inline espnet2/curriculum/curriculum_sampler.py:67]: small optimization

File: espnet2/curriculum/curriculum_sampler.py [hunk 3]
```diff
@@ -80,8 +81,7 @@ def __init__(
         for id in tasks:
             task_keys[tasks[id][0]].append(id)
             
-        # Sort samples in descending order
-        sorted_task_keys = [sorted(t, key=lambda k: first_utt2shape[k][0]) for t in task_keys]
+        sorted_task_keys = task_keys
```
Comment (ajhinsvark) [inline espnet2/curriculum/curriculum_sampler.py:84]: removed sorting since we shuffle them later anyways

File: espnet2/tasks/abs_task.py [hunk 4]
```diff
@@ -1502,18 +1502,32 @@ def build_curriculum_iter_factory(
...
+            split_tasks = []
+            for batches in tasks:
+                filtered_batches = 0
+                local_batches = []
+                for batch in batches:
+                    if len(batch) < world_size:
+                        filtered_batches += 1
+                        continue
+                    local_batches.append(batch[rank::world_size])
+                logging.warning(f"{filtered_batches} batches less than world size removed")
```
Comment (ajhinsvark) [inline espnet2/tasks/abs_task.py:1524]: I don't know if this is a problem since it removes batches with really long utterances.

File: espnet2/curriculum/curriculum_sampler.py (hunk 3 above)
Comment (ana-kuznetsova) [reply espnet2/curriculum/curriculum_sampler.py:84]: The training examples are being shuffled within each of k tasks. We still need overall sorting, because that's how we do the task split.

**Review by ana-kuznetsova — APPROVED** [review]
